### PR TITLE
kuro: init at 9.0.0

### PR DESCRIPTION
**Pull request by LostAttractor**
###### Description of changes
kuro: init at 9.0.0

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by LostAttractor** [comment]
Previously, because of some misunderstandings about the use input-methods on electron apps  on nixos, multiple force pushes were performed
Now this pull request is ready for review!

File: pkgs/applications/misc/kuro/default.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      categories = [ "Office" ];
+      startupWMClass = pname;
+    };
+
+  meta = with lib; {
+    description = ''
+      kuro is an unofficial, featureful, open source,
+      community-driven, free Microsoft To-Do app,
+      used by people in more than 120 countries.
+    '';
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:83]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md?plain=1#L40

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -2450,6 +2450,12 @@
     githubId = 89596;
     name = "Florian Friesdorf";
   };
+  ChaosAttractor = {
+    email = "lostattractor@gmail.com";
+    github = "LostAttractor";
+    githubId = 46527539;
+    name = "ChaosAttractor";
+  };
```
Comment (zendo) [inline maintainers/maintainer-list.nix:2458]: Don't need this right now, and you should squash the commits into one: `kuro: init at 8.1.8`

File: pkgs/applications/misc/kuro/default.nix [hunk 3]
```diff
@@ -0,0 +1,85 @@
+{ lib
+, fetchFromGitHub
+, makeWrapper
+, makeDesktopItem
+, mkYarnPackage
+, electron_22
+, yarn2nix
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:7]:
```suggestion
, makeDesktopItem
, copyDesktopItems
, mkYarnPackage
, electron
```

File: pkgs/applications/misc/kuro/default.nix [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+let
+  pname = "kuro";
+  executableName = pname;
+  version = "8.1.8";
+  electron = electron_22;
+
+in mkYarnPackage rec {
+  name = "${pname}-${version}";
+  inherit version;
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:18]:
```suggestion
mkYarnPackage rec {
  pname = "kuro";
  version = "8.1.8";
```

File: pkgs/applications/misc/kuro/default.nix [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-+4HgH8HBQ+KIN2u+3ZsPMORPl5C7BO+rXR0RsA79EGg=";
+  };
+
+  packageJSON = ./package.json;
+  yarnLock = ./yarn.lock;
+  yarnNix = ./yarn.nix;
+
+  nativeBuildInputs = [ makeWrapper ];
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:31]:
```suggestion
  nativeBuildInputs = [
    makeWrapper
    copyDesktopItems
  ];
```

File: pkgs/applications/misc/kuro/default.nix [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+  installPhase = ''
+    # resources
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:34]:
```suggestion
  installPhase = ''
    runHook preInstall

    # resources
```

File: pkgs/applications/misc/kuro/default.nix [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    echo "Installing icons..."
+    for size in 16x16 24x24 32x32 48x48 64x64 72x72 96x96 128x128 192x192 256x256 512x512 1024x1024; do
+      install -Dm644 ./deps/kuro/static/Icon.png $out/share/icons/hicolor/$size/apps/kuro.png
+    done
+
+    # desktop item
+    echo "Installing desktop item..."
+    mkdir -p "$out/share"
+    ln -s "${desktopItem}/share/applications" "$out/share/applications"
+
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:52]:
```suggestion
```

File: pkgs/applications/misc/kuro/default.nix [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    # executable wrapper
+    makeWrapper '${electron}/bin/electron' "$out/bin/${executableName}" \
+      --argv0 "kuro" \
+      --add-flags "$out/share/kuro/electron"
+  '';
+
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:58]:
```suggestion

    runHook postInstall
  '';
```

File: pkgs/applications/misc/kuro/default.nix [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  # Do not attempt generating a tarball for contents again.
+  # note: `doDist = false;` does not work.
+  distPhase = ''
+    true
+  '';
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:63]:
```suggestion
  distPhase = "true";
```

File: pkgs/applications/misc/kuro/default.nix [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      name = pname;
+      exec = executableName;
+      icon = pname;
+      desktopName = pname;
+      genericName = "Microsoft To-Do Client";
+      comment = concatStringsSep " "
+                  (splitString "\n" meta.description);
+      categories = [ "Office" ];
+      startupWMClass = pname;
+    };
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:76]:
```suggestion
  desktopItems = [
    (makeDesktopItem {
      name = pname;
      exec = pname;
      icon = pname;
      desktopName = "Kuro";
      genericName = "Microsoft To-Do Client";
      comment = meta.description;
      categories = [ "Office" ];
      startupWMClass = pname;
    })
  ];
```

File: pkgs/applications/misc/kuro/default.nix [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+  meta = with lib; {
+    description = "An unofficial, featureful, open source, community-driven, free Microsoft To-Do app.";
```
Comment (zendo) [inline pkgs/applications/misc/kuro/default.nix:79]:
```suggestion
    description = "An unofficial, featureful, open source, community-driven, free Microsoft To-Do app";
```

File: pkgs/top-level/all-packages.nix [hunk 12]
```diff
@@ -30482,6 +30482,8 @@ with pkgs;
 
   kupfer = callPackage ../applications/misc/kupfer { };
 
+  kuro = callPackage ../applications/misc/kuro { };
+
```
Comment (zendo) [inline pkgs/top-level/all-packages.nix:30486]:
```suggestion
  kuro = callPackage ../applications/misc/kuro {
    electron = electron_22;
  };
```

**Comment by LostAttractor** [comment]
Kuro has been updated to 9.0.0, So I have updated it to latest version.
This package ready for review.

File: pkgs/applications/misc/kuro/default.nix [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+  nativeBuildInputs = [
+    makeWrapper
+    copyDesktopItems
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    # resources
+    echo "Installing resources..."
```
Comment (ocfox) [inline pkgs/applications/misc/kuro/default.nix:35]:
```suggestion
```

File: pkgs/applications/misc/kuro/default.nix [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    mkdir -p "$out/share/kuro/electron"
+    cp -r './deps/kuro/static' "$out/share/kuro/electron"
+    cp -r './deps/kuro/src' "$out/share/kuro/electron"
+    cp -r './deps/kuro/index.js' "$out/share/kuro/electron"
+    cp -r './node_modules' "$out/share/kuro/electron"
+
+    # icons
+    echo "Installing icons..."
```
Comment (ocfox) [inline pkgs/applications/misc/kuro/default.nix:43]:
```suggestion
```

File: pkgs/applications/misc/kuro/default.nix [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+  installPhase = ''
+    runHook preInstall
+
+    # resources
+    mkdir -p "$out/share/kuro/electron"
+    cp -r './deps/kuro/static' "$out/share/kuro/electron"
+    cp -r './deps/kuro/src' "$out/share/kuro/electron"
+    cp -r './deps/kuro/index.js' "$out/share/kuro/electron"
+    cp -r './node_modules' "$out/share/kuro/electron"
```
Comment (NickCao) [inline pkgs/applications/misc/kuro/default.nix:39]: Can we just use `electron-builder`?

Comment (LostAttractor) [reply pkgs/applications/misc/kuro/default.nix]: Ok, I have switched to electron-builder now

File: pkgs/applications/misc/kuro/default.nix [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    runHook preInstall
+
+    # resources
+    mkdir -p "$out/share/lib/kuro"
+    cp -r ./deps/kuro/dist/*-unpacked/{locales,resources{,.pak}} "$out/share/lib/kuro"
+
+    # icons
+    for size in 16x16 24x24 32x32 48x48 64x64 72x72 96x96 128x128 192x192 256x256 512x512 1024x1024; do
+      install -Dm644 ./deps/kuro/static/Icon.png $out/share/icons/hicolor/$size/apps/kuro.png
+    done
```
Comment (NickCao) [inline pkgs/applications/misc/kuro/default.nix:54]: Pretty sure that a single icon file does not fit all sizes.

File: pkgs/applications/misc/kuro/default.nix [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+    # executable wrapper
+    makeWrapper '${electron}/bin/electron' "$out/bin/${executableName}" \
```
Comment (NickCao) [inline pkgs/applications/misc/kuro/default.nix:57]:
```suggestion
    makeWrapper '${electron}/bin/electron' "$out/bin/kuro" \
```

File: pkgs/applications/misc/kuro/default.nix [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+, makeDesktopItem
+, copyDesktopItems
+, mkYarnPackage
+, electron
+}:
+
+mkYarnPackage rec {
+  pname = "kuro";
+  version = "9.0.0";
+  executableName = pname;
```
Comment (NickCao) [inline pkgs/applications/misc/kuro/default.nix:13]:
```suggestion
```
A bit redundant.

**Review by NickCao — APPROVED** [review]
diff LGTM, waiting for ofborg.

**Pull request merged**